nav2_dynamic_params/src/dynamic_params_validator.cpp

```diff
@@ -0,0 +1,158 @@
+
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+#include <limits>
+#include "nav2_dynamic_params/dynamic_params_validator.hpp"
+
+namespace nav2_dynamic_params
+{
+
+DynamicParamsValidator::DynamicParamsValidator(rclcpp::Node::SharedPtr node)
+: node_(node)
+{
+  node_->register_param_change_callback(
+    std::bind(&DynamicParamsValidator::param_validation_callback, this, std::placeholders::_1));
+}
+
+rcl_interfaces::msg::SetParametersResult DynamicParamsValidator::param_validation_callback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+  auto result = rcl_interfaces::msg::SetParametersResult();
+  result.successful = true;
+
+  for (auto parameter : parameters) {
+    RCLCPP_INFO(node_->get_logger(),
+      "Parameter Change Request: %s", (parameter.get_name()).c_str());
+
+    // Validate Parameter Type
+    for (std::map<std::string, rclcpp::ParameterType>::iterator it = param_map_.begin();
+      it != param_map_.end(); ++it)
+    {
+      if (validate_param(it->first, parameter, it->second)) {
+        result.successful &= true;
+      } else {
+        result.successful = false;
+      }
+    }
+
+    // Validate Parameter Bounds
+    for (std::map<std::string, std::pair<double, double>>::iterator it = param_bound_map_.begin();
+      it != param_bound_map_.end(); ++it)
+    {
+      if (validate_param_bounds(it->first, parameter, it->second)) {
+        result.successful &= true;
+      } else {
+        result.successful = false;
+      }
+    }
+  }
+  // Check Custom Validation Callback if provided
+  if (validation_callback_) {
+    result.successful &= (validation_callback_(parameters)).successful;
+  }
+
+  return result;
+}
+
+void DynamicParamsValidator::add_param(std::string param_name, rclcpp::ParameterType type)
+{
+  param_map_[param_name] = type;
+}
+
+void DynamicParamsValidator::add_param(
+  std::string param_name, rclcpp::ParameterType type,
+  std::pair<double, double> bounds)
+{
+  param_map_[param_name] = type;
+  param_bound_map_[param_name] = bounds;
+}
+
+void DynamicParamsValidator::add_param(
+  std::string param_name, rclcpp::ParameterType type,
+  std::pair<double, double> bounds, int ignore_bound)
+{
+  param_map_[param_name] = type;
+  if (ignore_bound) {
+    bounds.second = std::numeric_limits<double>::infinity();
+  } else {
+    bounds.first = -std::numeric_limits<double>::infinity();
+  }
+  param_bound_map_[param_name] = bounds;
+}
+
+void DynamicParamsValidator::add_param(std::map<std::string, rclcpp::ParameterType> map)
+{
+  param_map_.insert(map.begin(), map.end());
+}
+
+void DynamicParamsValidator::set_validation_callback(
+  std::function<rcl_interfaces::msg::SetParametersResult(
+    const std::vector<rclcpp::Parameter> &)> callback)
+{
+  validation_callback_ = callback;
+}
+
+bool DynamicParamsValidator::validate_param(
+  std::string param_name, rclcpp::Parameter param,
+  rclcpp::ParameterType type)
+{
+  if (param.get_name() == param_name) {
```

Comment:
If you use the map to look up the type by name, then you can skip this param_name check.